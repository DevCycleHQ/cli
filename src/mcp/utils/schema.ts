import { z, ZodTypeAny, ZodRawShape } from 'zod'
import type { Tool, ToolAnnotations } from '@modelcontextprotocol/sdk/types.js'
import { zodToJsonSchema } from 'zod-to-json-schema'

type JsonSchema = Record<string, unknown>

type ToolConfig = {
    description: string
    annotations?: ToolAnnotations
    inputSchema?: unknown
    outputSchema?: unknown
}

const ENABLE_OUTPUT_SCHEMAS = process.env.ENABLE_OUTPUT_SCHEMAS === 'true'

function isZodType(value: unknown): value is ZodTypeAny {
    return !!value && typeof value === 'object' && 'safeParse' in value
}

function maybeWrapShape(schema: unknown): ZodTypeAny | undefined {
    if (!schema || typeof schema !== 'object') return undefined
    // If it's already a Zod type, leave it
    if (isZodType(schema)) return schema as ZodTypeAny
    // Be permissive: treat any plain object as a Zod object shape. This covers
    // passing `SomeSchema.shape` as well as inline shapes.
    try {
        return z.object(schema as ZodRawShape)
    } catch {
        return undefined
    }
}

export function toJsonSchema(
    schema: unknown,
    name?: string,
): JsonSchema | undefined {
    if (!schema) return undefined
    const zodSchema = isZodType(schema) ? schema : maybeWrapShape(schema)
    if (!zodSchema) return undefined

    const json = zodToJsonSchema(
        zodSchema,
        name ? { name } : undefined,
    ) as JsonSchema

    // Normalize the schema for MCP consumers that require `items` on arrays
    // Some generated schemas (e.g., z.array(z.any())) omit `items`. Add a minimal
    // empty schema so tools validate correctly.
    function ensureArrayItems(node: unknown): void {
        if (!node || typeof node !== 'object') return
        const obj = node as Record<string, unknown>
        if (obj.type === 'array' && obj.items === undefined) {
            obj.items = {}
        }
        for (const value of Object.values(obj)) {
            ensureArrayItems(value)
        }
    }
    ensureArrayItems(json)
    return json
}

type ToolRegistrationConfig = Omit<Tool, 'name'>

const defaultObjectSchema: JsonSchema = { type: 'object', properties: {} }

export function processToolConfig(
    name: string,
    config: ToolConfig,
): ToolRegistrationConfig {
    const processed: ToolRegistrationConfig = {
        description: config.description,
        annotations: config.annotations,
        inputSchema: defaultObjectSchema,
    }

    const inputJson = toJsonSchema(config.inputSchema, `${name}_input`)
    if (inputJson) processed.inputSchema = inputJson

    if (ENABLE_OUTPUT_SCHEMAS) {
        const outputJson = toJsonSchema(config.outputSchema, `${name}_output`)
        if (outputJson) processed.outputSchema = outputJson
    }

    return processed
}
