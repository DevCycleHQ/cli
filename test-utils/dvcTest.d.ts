export declare const dvcTest: () => import("fancy-test/lib/types").Base<{
    nock: number;
} & import("fancy-test/lib/types").Context, {
    skip: {
        output: unknown;
        args: [];
    };
} & {
    only: {
        output: unknown;
        args: [];
    };
} & {
    retries: {
        output: unknown;
        args: [count: number];
    };
} & {
    catch: {
        output: {
            error: Error;
        };
        args: [arg: string | RegExp | ((err: Error) => any), opts?: {
            raiseIfNotThrown?: boolean | undefined;
        } | undefined];
    };
} & {
    env: {
        output: unknown;
        args: [env: {
            [k: string]: string | null | undefined;
        }, opts?: import("fancy-test/lib/types").EnvOptions | undefined];
    };
} & {
    stub: {
        output: {
            stubs: any[];
        };
        args: [object: any, path: any, value: () => any];
    };
} & {
    stdin: {
        output: unknown;
        args: [input: string, delay?: number | undefined];
    };
} & {
    stderr: {
        output: {
            readonly stderr: string;
        };
        args: [opts?: {
            print?: boolean | undefined;
            stripColor?: boolean | undefined;
        } | undefined];
    };
} & {
    stdout: {
        output: {
            readonly stdout: string;
        };
        args: [opts?: {
            print?: boolean | undefined;
            stripColor?: boolean | undefined;
        } | undefined];
    };
} & {
    nock: {
        output: {
            nock: number;
        };
        args: [host: string, options: import("fancy-test/lib/types").NockOptions | import("fancy-test/lib/types").NockCallback, cb?: import("fancy-test/lib/types").NockCallback | undefined];
    };
} & {
    timeout: {
        output: {
            timeout: number;
        };
        args: [timeout?: number | undefined];
    };
} & {
    loadConfig: {
        output: {
            config: import("@oclif/core/lib/interfaces").Config;
        };
        args: [opts?: import("@oclif/test/lib/load-config").loadConfig.Options | undefined];
    };
} & {
    command: {
        output: {
            config: import("@oclif/core/lib/interfaces").Config;
            expectation: string;
            returned: unknown;
        };
        args: [args: string | string[], opts?: import("@oclif/test/lib/load-config").loadConfig.Options | undefined];
    };
} & {
    exit: {
        output: {
            error: any;
        };
        args: [code?: number | undefined];
    };
} & {
    hook: {
        output: {
            config: import("@oclif/core/lib/interfaces").Config;
            expectation: string;
            returned: unknown;
        };
        args: [event: string, hookOpts?: Record<string, unknown> | undefined, options?: import("@oclif/test/lib/load-config").loadConfig.Options | undefined];
    };
}>;
export declare const mockFeatures: ({
    key: string;
    name: string;
    _id: string;
    _project: string;
    source: string;
    _createdBy: string;
    createdAt: string;
    updatedAt: string;
    variations: never[];
    variables: never[];
    tags: never[];
    configurations: never[];
    sdkVisibility: {
        mobile: boolean;
        client: boolean;
        server: boolean;
    };
    settings: {
        optInEnabled?: undefined;
        publicName?: undefined;
        publicDescription?: undefined;
    };
    readonly: boolean;
    controlVariation: string;
    status?: undefined;
    type?: undefined;
    description?: undefined;
} | {
    key: string;
    name: string;
    _id: string;
    _project: string;
    source: string;
    status: string;
    type: string;
    description: string;
    _createdBy: string;
    createdAt: string;
    updatedAt: string;
    variations: {
        _id: string;
        key: string;
        name: string;
        variables: {
            'togglebot-speed': string;
            'example-text': string;
            'togglebot-wink': boolean;
        };
    }[];
    controlVariation: string;
    variables: {
        _id: string;
        _project: string;
        _feature: string;
        name: string;
        key: string;
        type: string;
        status: string;
        source: string;
        _createdBy: string;
        createdAt: string;
        updatedAt: string;
    }[];
    tags: never[];
    readonly: boolean;
    settings: {
        optInEnabled: boolean;
        publicName: string;
        publicDescription: string;
    };
    sdkVisibility: {
        mobile: boolean;
        client: boolean;
        server: boolean;
    };
    configurations: {
        _feature: string;
        _environment: string;
        _createdBy: string;
        status: string;
        startedAt: string;
        updatedAt: string;
        targets: {
            _id: string;
            name: string;
            audience: {
                name: string;
                filters: {
                    operator: string;
                    filters: {
                        type: string;
                    }[];
                };
            };
            distribution: {
                _variation: string;
                percentage: number;
            }[];
        }[];
        readonly: boolean;
    }[];
})[];
//# sourceMappingURL=dvcTest.d.ts.map