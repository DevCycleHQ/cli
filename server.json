{
    "$schema": "https://static.modelcontextprotocol.io/schemas/2025-07-09/server.schema.json",
    "name": "com.devcycle/mcp",
    "description": "DevCycle MCP server for feature flag management",
    "version": "6.1.2",
    "repository": {
        "url": "https://github.com/DevCycleHQ/cli",
        "source": "github"
    },
    "website_url": "https://docs.devcycle.com/cli-mcp/mcp-getting-started",
    "remotes": [
        {
            "type": "streamable-http",
            "url": "https://mcp.devcycle.com/mcp"
        },
        {
            "type": "sse",
            "url": "https://mcp.devcycle.com/sse"
        }
    ]
}
