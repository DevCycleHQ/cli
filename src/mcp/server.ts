import { Server } from '@modelcontextprotocol/sdk/server/index.js'
import {
    CallToolRequestSchema,
    ListToolsRequestSchema,
    Tool,
} from '@modelcontextprotocol/sdk/types.js'
import { DevCycleAuth } from './utils/auth'
import { DevCycleApiClient } from './utils/api'
import Writer from '../ui/writer'
import {
    featureToolDefinitions,
    featureToolHandlers,
} from './tools/featureTools'
import {
    environmentToolDefinitions,
    environmentToolHandlers,
} from './tools/environmentTools'
import {
    projectToolDefinitions,
    projectToolHandlers,
} from './tools/projectTools'
import {
    variableToolDefinitions,
    variableToolHandlers,
} from './tools/variableTools'
import {
    selfTargetingToolDefinitions,
    selfTargetingToolHandlers,
} from './tools/selfTargetingTools'

// Environment variable to control output schema inclusion
const ENABLE_OUTPUT_SCHEMAS = process.env.ENABLE_OUTPUT_SCHEMAS === 'true'
if (ENABLE_OUTPUT_SCHEMAS) {
    console.error('DevCycle MCP Server - Output Schemas: ENABLED')
}

const ENABLE_DVC_MCP_DEBUG = process.env.ENABLE_DVC_MCP_DEBUG === 'true'

// Tool handler function type
export type ToolHandler = (
    args: unknown,
    apiClient: DevCycleApiClient,
) => Promise<any>

// Function to conditionally remove outputSchema from tool definitions
const processToolDefinitions = (tools: Tool[]): Tool[] => {
    if (ENABLE_OUTPUT_SCHEMAS) {
        return tools
    }

    // Remove outputSchema from all tools when disabled
    return tools.map((tool) => {
        const { outputSchema, ...toolWithoutSchema } = tool
        return toolWithoutSchema
    })
}

// Combine all tool definitions
const allToolDefinitions: Tool[] = processToolDefinitions([
    ...featureToolDefinitions,
    ...environmentToolDefinitions,
    ...projectToolDefinitions,
    ...variableToolDefinitions,
    ...selfTargetingToolDefinitions,
])

// Combine all tool handlers
const allToolHandlers: Record<string, ToolHandler> = {
    ...featureToolHandlers,
    ...environmentToolHandlers,
    ...projectToolHandlers,
    ...variableToolHandlers,
    ...selfTargetingToolHandlers,
}

export class DevCycleMCPServer {
    private auth: DevCycleAuth
    private apiClient: DevCycleApiClient
    private writer: Writer

    constructor(private server: Server) {
        this.writer = new Writer()
        this.writer.headless = true // Always headless for MCP
        this.auth = new DevCycleAuth()
        this.apiClient = new DevCycleApiClient(this.auth)
    }

    async initialize() {
        try {
            await this.setupAuth()
            this.setupToolHandlers()
            this.setupErrorHandling()
        } catch (error) {
            console.error('Failed to initialize MCP server:', error)
            throw error
        }
    }

    private async setupAuth() {
        try {
            await this.auth.initialize()
        } catch (error) {
            console.error('Failed to initialize authentication:', error)
            throw error
        }
    }

    private handleToolError(error: unknown, toolName: string) {
        console.error(`Error in tool handler ${toolName}:`, error)

        let errorMessage = 'Unknown error'
        let errorType = 'UNKNOWN_ERROR'
        let suggestions: string[] = []

        if (error instanceof Error) {
            errorMessage = error.message
            errorType = this.categorizeError(error.message)
            suggestions = this.getErrorSuggestions(errorType)
        } else if (error && typeof error === 'string') {
            errorMessage = error
        } else if (error && typeof error === 'object') {
            errorMessage = JSON.stringify(error)
        }

        const errorResponse = {
            error: true,
            type: errorType,
            message: errorMessage,
            tool: toolName,
            suggestions,
            timestamp: new Date().toISOString(),
        }

        return {
            content: [
                {
                    type: 'text',
                    text: JSON.stringify(errorResponse, null, 2),
                },
            ],
        }
    }

    private categorizeError(errorMessage: string): string {
        const lowerMessage = errorMessage.toLowerCase()

        switch (true) {
            case lowerMessage.includes('zodios: invalid response') ||
                lowerMessage.includes('invalid_type') ||
                lowerMessage.includes('expected object, received'):
                return 'SCHEMA_VALIDATION_ERROR'

            case lowerMessage.includes('401') ||
                lowerMessage.includes('unauthorized'):
                return 'AUTHENTICATION_ERROR'

            case lowerMessage.includes('403') ||
                lowerMessage.includes('forbidden'):
                return 'PERMISSION_ERROR'

            case lowerMessage.includes('404') ||
                lowerMessage.includes('not found'):
                return 'RESOURCE_NOT_FOUND'

            case lowerMessage.includes('400') ||
                lowerMessage.includes('bad request'):
                return 'VALIDATION_ERROR'

            case lowerMessage.includes('429') ||
                lowerMessage.includes('rate limit'):
                return 'RATE_LIMIT_ERROR'

            case lowerMessage.includes('enotfound') ||
                lowerMessage.includes('network'):
                return 'NETWORK_ERROR'

            case lowerMessage.includes('project') &&
                lowerMessage.includes('not found'):
                return 'PROJECT_ERROR'

            default:
                return 'UNKNOWN_ERROR'
        }
    }

    private getErrorSuggestions(errorType: string): string[] {
        switch (errorType) {
            case 'SCHEMA_VALIDATION_ERROR':
                return [
                    'The API response format has changed or is unexpected',
                    'This may be a temporary API issue - try again in a moment',
                    'Contact DevCycle support if the issue persists',
                ]

            case 'AUTHENTICATION_ERROR':
                return [
                    'Run "dvc login sso" to re-authenticate the devcycle cli',
                    'Verify your API credentials are correct',
                    'Check if your token has expired',
                ]

            case 'PERMISSION_ERROR':
                return [
                    'Verify your account has permissions for this operation',
                    'Check if you have access to the selected project',
                    'Contact your DevCycle admin for permissions',
                ]

            case 'RESOURCE_NOT_FOUND':
                return [
                    'Verify the resource key/ID is correct',
                    'Check if the resource exists in the selected project',
                    "Ensure you're in the correct environment",
                ]

            case 'VALIDATION_ERROR':
                return [
                    'Check the provided parameters are valid',
                    'Verify required fields are not missing',
                    'Review parameter format and constraints',
                ]

            case 'RATE_LIMIT_ERROR':
                return [
                    'Wait a moment before trying again',
                    'Consider reducing the frequency of requests',
                ]

            case 'NETWORK_ERROR':
                return [
                    'Check your internet connection',
                    'Verify firewall settings allow DevCycle API access',
                    'Try again in a few moments',
                ]

            case 'PROJECT_ERROR':
                return [
                    'Run "dvc projects select" to choose a valid project',
                    'Verify the project key is correct',
                    'Check if you have access to this project',
                ]

            default:
                return []
        }
    }

    private setupToolHandlers() {
        this.server.setRequestHandler(ListToolsRequestSchema, async () => ({
            tools: allToolDefinitions,
        }))

        this.server.setRequestHandler(
            CallToolRequestSchema,
            async (request) => {
                const { name, arguments: args } = request.params

                try {
                    const handler = allToolHandlers[name]
                    if (!handler) {
                        throw new Error(`Unknown tool: ${name}`)
                    }

                    const result = await handler(args, this.apiClient)

                    // Return structured content only if output schemas are enabled
                    if (ENABLE_OUTPUT_SCHEMAS) {
                        // Check if tool has output schema
                        const toolDef = allToolDefinitions.find(
                            (tool) => tool.name === name,
                        )

                        if (toolDef?.outputSchema) {
                            // For tools with output schemas, return structured JSON content
                            const mcpResult = {
                                content: [
                                    {
                                        type: 'json',
                                        json: result,
                                    },
                                ],
                            }
                            if (ENABLE_DVC_MCP_DEBUG) {
                                console.error(
                                    `MCP ${name} structured JSON result:`,
                                    JSON.stringify(mcpResult, null, 2),
                                )
                            }
                            return mcpResult
                        }
                    }

                    // Default: return as text content (for disabled schemas or tools without schemas)
                    const mcpResult = {
                        content: [
                            {
                                type: 'text',
                                text: JSON.stringify(result, null, 2),
                            },
                        ],
                    }
                    if (ENABLE_DVC_MCP_DEBUG) {
                        console.error(
                            `MCP ${name} text result:`,
                            JSON.stringify(mcpResult, null, 2),
                        )
                    }
                    return mcpResult
                } catch (error) {
                    return this.handleToolError(error, name)
                }
            },
        )
    }

    private setupErrorHandling() {
        this.server.onerror = (error: Error) => {
            console.error('MCP Server Error:', error)
        }
    }
}
