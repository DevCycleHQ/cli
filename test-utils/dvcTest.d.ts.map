{"version":3,"file":"dvcTest.d.ts","sourceRoot":"","sources":["dvcTest.ts"],"names":[],"mappings":"AAIA,eAAO,MAAM,OAAO;;;;;;;;;;;;;;;;;;;;;;iEAKqS,CAAC;4BAAgC,CAAC;;;;;;;;eAAyL,CAAC;;;;;;;;;;;;mCAAsQ,CAAC;;;;;;;mBAAuI,CAAC;iBAAqB,CAAC;sBAA6C,CAAC;;;;;;;;mBAG18B,CAAA;iBACd,CAAC;sBACD,CAAF;;;;;;;;kIAQE,CAAZ;;;;;;;sBAQa,CAAC;;;;;;;mBAKP,CAAC;;;;;;;;;4CAWJ,CAAF;;;;;;;mBAMI,CAAC;;;;;;;;;sCAOG,CAAC,8CAEK,CAAC;;EArCX,CAAA;AAEV,eAAO,MAAM,YAAY;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;;IAwKxB,CAAA"}