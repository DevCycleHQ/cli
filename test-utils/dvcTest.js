"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.mockFeatures = exports.dvcTest = void 0;
/* eslint-disable max-len */
const test_1 = require("@oclif/test");
const common_1 = require("../src/api/common");
const dvcTest = () => test_1.test
    .nock(common_1.AUTH_URL, (api) => {
    api.post('/oauth/token').reply(200, {
        access_token: 'eyJhbGciOiJSUzI1NiIsInR5cCI6IkpXVCIsImtpZCI6IlhpTzk1Xzllbk53Z1NNSkZRSXZNUiJ9.eyJodHRwczovL2RldmN5Y2xlLmNvbS9lbWFpbCI6InRlc3RAdGFwbHl0aWNzLmNvbSIsImh0dHBzOi8vZGV2Y3ljbGUuY29tL2lzR2VuZXJpY0RvbWFpbiI6ZmFsc2UsImh0dHBzOi8vZGV2Y3ljbGUuY29tL2FsbG93T3JnRGlzY292ZXJ5Ijp0cnVlLCJpc3MiOiJodHRwczovL2F1dGguZGV2Y3ljbGUuY29tLyIsInN1YiI6Imdvb2dsZS1vYXV0aDJ8MTExNTU5MDA2NTYzMzMzMzM0MjE0IiwiYXVkIjoiaHR0cHM6Ly9hcGkuZGV2Y3ljbGUuY29tLyIsImlhdCI6MTY4Nzk2NjU0MiwiZXhwIjo5OTk5OTk5OTk5LCJhenAiOiJFdjlKMERHeFIzS2hyS2Fad1k2amxjY21qbDdKR0tFWCIsIm9yZ19pZCI6Im9yZ19VOUY4WU1hVENoVEVuZFd3IiwicGVybWlzc2lvbnMiOlsiY3JlYXRlOnJlc291cmNlcyIsImRlbGV0ZTpyZXNvdXJjZXMiLCJyZWFkOmN1cnJlbnRfb3JnYW5pemF0aW9uIiwicmVhZDpyZXNvdXJjZXMiLCJyZWFkOnVzZXJzIiwidXBkYXRlOnJlc291cmNlcyJdfQ==.bd8rLSYh6ZtnzEgG-Mya86wnOuZcRC97tb4_8gdjTgH2aUCkNRDKcJvLkFL_wYszBurFs75n-BdHKnv9H8yy23fdQbPxX-5Wjq1vvS5PN3DdpvGtc5EuumFBxgwUO3WTmId2znjbMBWIVOM5bViRN8Pba4X6XZSSQ6hxd7-30Txj_5dOsPNikZHjBDuTCqlQxdrZu_ER0JKKkRwdlR8h7qsGMBrTeVqBhIFQXQIFns440FZsCGaOK4hFw9vg6remaBu3HTQRWClXHVwh03lwFTX8MTfwRwFDGhl988zbzK38AAncoKcXKCMrG2GXVswar0RaQWhLSL9eLuX0PI99Ww',
    });
})
    .nock(common_1.BASE_URL, (api) => {
    api.get('/v1/projects').reply(200, [
        {
            key: 'project',
            settings: {
                obfuscation: {},
            },
        },
        {
            key: 'test-project',
        },
    ]);
});
exports.dvcTest = dvcTest;
exports.mockFeatures = [
    {
        key: 'feature-1',
        name: 'Feature 1',
        _id: '61450f3daec96f5cf4a49946',
        _project: 'test-project',
        source: 'api',
        _createdBy: 'test-user',
        createdAt: '2021-09-15T12:00:00Z',
        updatedAt: '2021-09-15T12:00:00Z',
        variations: [],
        variables: [],
        tags: [],
        configurations: [],
        sdkVisibility: {
            mobile: true,
            client: true,
            server: true,
        },
        settings: {},
        readonly: false,
        controlVariation: 'control',
    },
    {
        key: 'feature-2',
        name: 'Feature 2',
        _id: '61450f3daec96f5cf4a49947',
        _project: 'test-project',
        source: 'api',
        _createdBy: 'test-user',
        createdAt: '2021-09-15T12:00:00Z',
        updatedAt: '2021-09-15T12:00:00Z',
        variations: [],
        variables: [],
        tags: [],
        configurations: [],
        sdkVisibility: {
            mobile: true,
            client: true,
            server: true,
        },
        settings: {},
        readonly: false,
        controlVariation: 'control',
    },
    {
        key: 'feature-with-optin',
        name: 'feature with optIn',
        _id: '68220a9dbc3cb4d3d14ba7bf',
        _project: '6819062b8affcfbc2be4d4d5',
        source: 'dashboard',
        status: 'active',
        type: 'release',
        description: 'description',
        _createdBy: 'google-oauth2|112025847388655732143',
        createdAt: '2025-05-12T14:50:05.620Z',
        updatedAt: '2025-05-21T15:06:50.696Z',
        variations: [
            {
                _id: '68220a9dbc3cb4d3d14ba7d4',
                key: 'variation-base',
                name: 'Base Variation',
                variables: {
                    'togglebot-speed': 'off',
                    'example-text': 'step-1',
                    'togglebot-wink': false,
                },
            },
            {
                _id: '68220a9dbc3cb4d3d14ba7d5',
                key: 'variation-spin',
                name: 'Spin Variation',
                variables: {
                    'togglebot-speed': 'slow',
                    'example-text': 'step-2',
                    'togglebot-wink': true,
                },
            },
        ],
        controlVariation: 'variation-base',
        variables: [
            {
                _id: '68220a9dbc3cb4d3d14ba7c5',
                _project: '6819062b8affcfbc2be4d4d5',
                _feature: '68220a9dbc3cb4d3d14ba7bf',
                name: 'Togglebot Spin Speed',
                key: 'togglebot-speed',
                type: 'String',
                status: 'active',
                source: 'dashboard',
                _createdBy: 'google-oauth2|112025847388655732143',
                createdAt: '2025-05-12T14:50:05.635Z',
                updatedAt: '2025-05-12T14:50:05.635Z',
            },
            {
                _id: '68220a9dbc3cb4d3d14ba7c6',
                _project: '6819062b8affcfbc2be4d4d5',
                _feature: '68220a9dbc3cb4d3d14ba7bf',
                name: 'Togglebot Wink Enabled',
                key: 'togglebot-wink',
                type: 'Boolean',
                status: 'active',
                source: 'dashboard',
                _createdBy: 'google-oauth2|112025847388655732143',
                createdAt: '2025-05-12T14:50:05.635Z',
                updatedAt: '2025-05-12T14:50:05.635Z',
            },
            {
                _id: '68220a9dbc3cb4d3d14ba7c7',
                _project: '6819062b8affcfbc2be4d4d5',
                _feature: '68220a9dbc3cb4d3d14ba7bf',
                name: 'Example App Display Text',
                key: 'example-text',
                type: 'String',
                status: 'active',
                source: 'dashboard',
                _createdBy: 'google-oauth2|112025847388655732143',
                createdAt: '2025-05-12T14:50:05.635Z',
                updatedAt: '2025-05-12T14:50:05.635Z',
            },
        ],
        tags: [],
        readonly: false,
        settings: {
            optInEnabled: true,
            publicName: 'public feature name',
            publicDescription: 'public feature description',
        },
        sdkVisibility: {
            mobile: true,
            client: true,
            server: true,
        },
        configurations: [
            {
                _feature: '68220a9dbc3cb4d3d14ba7bf',
                _environment: '6819062b8affcfbc2be4d4d8',
                _createdBy: 'google-oauth2|112025847388655732143',
                status: 'active',
                startedAt: '2025-05-12T14:50:05.815Z',
                updatedAt: '2025-05-21T15:04:01.135Z',
                targets: [
                    {
                        _id: '682380cf606a1643f8bd01e3',
                        name: 'Opt-In Users',
                        audience: {
                            name: 'Opt-In Users',
                            filters: {
                                operator: 'and',
                                filters: [
                                    {
                                        type: 'optIn',
                                    },
                                ],
                            },
                        },
                        distribution: [
                            {
                                _variation: '68220a9dbc3cb4d3d14ba7d4',
                                percentage: 1,
                            },
                        ],
                    },
                ],
                readonly: false,
            },
        ],
    },
];
//# sourceMappingURL=dvcTest.js.map