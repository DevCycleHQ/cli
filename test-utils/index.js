"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
tslib_1.__exportStar(require("./dvcTest"), exports);
tslib_1.__exportStar(require("./setCurrentTestFile"), exports);
//# sourceMappingURL=index.js.map