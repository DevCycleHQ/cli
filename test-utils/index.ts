export * from "./dvcTest";
export * from "./setCurrentTestFile";
