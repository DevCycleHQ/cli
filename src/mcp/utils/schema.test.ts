import { expect } from 'chai'
import * as chai from 'chai'
import { jestSnapshotPlugin } from 'mocha-chai-jest-snapshot'
import { setCurrentTestFile } from '../../../test-utils'
import sinon from 'sinon'
import { z } from 'zod'
import { processToolConfig, toJsonSchema } from './schema'
import { DevCycleMCPServer } from '../server'

function expectHasObjectSchema(json: any, defName?: string) {
    if (json && json.type === 'object') {
        expect(json).to.have.property('properties')
        return
    }
    expect(json).to.have.property('$ref')
    expect(json).to.have.property('definitions')
    const name = defName || ((json.$ref as string).split('/').pop() as string)
    expect(json.definitions).to.have.property(name)
    const def = json.definitions[name]
    expect(def).to.have.property('type', 'object')
    expect(def).to.have.property('properties')
}

describe('schema conversion', () => {
    beforeEach(setCurrentTestFile(__filename))
    chai.use(jestSnapshotPlugin())
    it('toJsonSchema converts a Zod object schema', () => {
        const Sample = z
            .object({
                name: z.string().describe('n'),
                page: z.number().min(1).default(1).describe('p'),
            })
            .describe('root')
        const json = toJsonSchema(Sample, 'Sample') as any
        expect(json).to.be.an('object')
        expectHasObjectSchema(json, 'Sample')
        expect(JSON.stringify(json, null, 2)).toMatchSnapshot()
    })

    it('processToolConfig converts a Zod shape object', () => {
        const Sample = z.object({ key: z.string().describe('k') })
        const config = processToolConfig('shape_test', {
            description: 'd',
            inputSchema: Sample.shape,
        }) as any
        expect(config).to.have.property('inputSchema')
        expectHasObjectSchema(config.inputSchema)
        expect(JSON.stringify(config.inputSchema, null, 2)).toMatchSnapshot()
    })

    it('registerToolWithErrorHandling uses JSON Schema input', () => {
        const server = { registerTool: sinon.stub() } as any
        const mcp = new DevCycleMCPServer(server)
        const Args = z.object({ feature: z.string() })
        mcp.registerToolWithErrorHandling(
            'tool_a',
            {
                description: 'd',
                inputSchema: Args.shape,
            },
            async () => ({}),
        )

        const call = (server.registerTool as sinon.SinonStub).getCall(0)
        expect(call).to.exist
        const cfg = call.args[1]
        expect(cfg).to.have.property('inputSchema')
        expectHasObjectSchema(cfg.inputSchema)
        expect(JSON.stringify(cfg.inputSchema, null, 2)).toMatchSnapshot()
    })
})
