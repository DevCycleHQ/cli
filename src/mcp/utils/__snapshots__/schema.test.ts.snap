// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schema conversion processToolConfig converts a Zod shape object 1`] = `
"{
  \\"$ref\\": \\"#/definitions/shape_test_input\\",
  \\"definitions\\": {
    \\"shape_test_input\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"key\\": {
          \\"type\\": \\"string\\",
          \\"description\\": \\"k\\"
        }
      },
      \\"required\\": [
        \\"key\\"
      ],
      \\"additionalProperties\\": false
    }
  },
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\"
}"
`;

exports[`schema conversion registerToolWithErrorHandling uses JSON Schema input 1`] = `
"{
  \\"$ref\\": \\"#/definitions/tool_a_input\\",
  \\"definitions\\": {
    \\"tool_a_input\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"feature\\": {
          \\"type\\": \\"string\\"
        }
      },
      \\"required\\": [
        \\"feature\\"
      ],
      \\"additionalProperties\\": false
    }
  },
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\"
}"
`;

exports[`schema conversion toJsonSchema converts a Zod object schema 1`] = `
"{
  \\"$ref\\": \\"#/definitions/Sample\\",
  \\"definitions\\": {
    \\"Sample\\": {
      \\"type\\": \\"object\\",
      \\"properties\\": {
        \\"name\\": {
          \\"type\\": \\"string\\",
          \\"description\\": \\"n\\"
        },
        \\"page\\": {
          \\"type\\": \\"number\\",
          \\"minimum\\": 1,
          \\"default\\": 1,
          \\"description\\": \\"p\\"
        }
      },
      \\"required\\": [
        \\"name\\"
      ],
      \\"additionalProperties\\": false,
      \\"description\\": \\"root\\"
    }
  },
  \\"$schema\\": \\"http://json-schema.org/draft-07/schema#\\"
}"
`;
